aten/src/ATen/core/ivalue.cpp

```diff
@@ -604,4 +604,43 @@ getClassConverter() {
       classConverter;
   return classConverter;
 }
+
+CAFFE2_API intrusive_ptr<ivalue::Future> collectAll(
+    List<intrusive_ptr<ivalue::Future>> srcs) {
+  struct Ctx {
+    explicit Ctx(List<intrusive_ptr<ivalue::Future>> srcs)
+        : remaining(srcs.size()),
+          srcFutures(std::move(srcs)),
+          asIvalue(srcFutures),
+          dstFuture(make_intrusive<ivalue::Future>(asIvalue.type())) {}
+    std::atomic<int32_t> remaining{0};
+    List<intrusive_ptr<ivalue::Future>> srcFutures;
+    IValue asIvalue;
+    intrusive_ptr<ivalue::Future> dstFuture;
+  };
+
+  auto ctx = std::make_shared<Ctx>(std::move(srcs));
+  std::function<void()> func = [ctx]() {
+    if (--ctx->remaining == 0) {
+      ctx->dstFuture->markCompleted(ctx->asIvalue);
+    }
+  };
+  for (int32_t tot = ctx->srcFutures.size(), i = 0; i < tot; ++i) {
+    ctx->srcFutures.get(i)->addCallback(func);
+  }
+  if (ctx->srcFutures.size() == 0) {
+    ctx->dstFuture->markCompleted(ctx->asIvalue);
+  }
+  return ctx->dstFuture;
+}
+
+CAFFE2_API intrusive_ptr<ivalue::Future> collectAll(
+    std::vector<intrusive_ptr<ivalue::Future>> srcs) {
+  List<intrusive_ptr<ivalue::Future>> asList(FutureType::create(AnyType::get()));
```

Comment:
we should create the List of Future with type as the type inside `srcs` i think? e.g. FutureType::create(srcs[0].type())

Reply:
How is that better than this approach? Is it set in stone that all Future target types must be the same?

Reply:
Yes, this also make sure the return type remains accurate according to its input future_list, so that adding torchscript support for this API is valid that we have accurate type information from the output of this API.

Reply:
Sure, that's fine.
 - The `c10::List` version of the api was already type-preserving, so no issues there.
 - Updating the vector version to use the appropriate type if they all match.
 - In the process, I found that `ivalue::Future`'s handling of typing is a bit inconsistent - `Future:type()` was like `List::elementType()` than e.g. the `Tuple::type()`. I ended up renaming this, and we can introduce a separate `Future::type()` if necessary.